lib/puma/server.rb

```diff
@@ -320,41 +320,65 @@ def handle_servers
-        when :proxy_protocol
-          [:expect_proxy_proto=, @options[:remote_address_proxy_protocol]]
-        else
-          [nil, nil]
-        end
-
-        while @status == :run || (drain && shutting_down?)
+        # option to local variables
+        # Using local variables avoid calling `@options#[]` method each time we need the value
+        remote_address_option = @options[:remote_address]
+        remote_address_value_option = @options[:remote_address_value]
+        remote_address_header_option = @options[:remote_address_header]
+        remote_address_proxy_protocol_option = @options[:remote_address_proxy_protocol]
+        wait_for_less_busy_worker = @options[:wait_for_less_busy_worker]
+        wait_for_less_busy_worker_enabled = !@options[:wait_for_less_busy_worker].nil?
+
+        is_shutting_down = false
+        while @status == :run || (drain && (is_shutting_down = shutting_down?))
           begin
-            ios = IO.select sockets, nil, nil, (shutting_down? ? 0 : nil)
+            ios = IO.select(sockets, nil, nil, (is_shutting_down ? 0 : nil))
             break unless ios
             ios.first.each do |sock|
               if sock == check
                 break if handle_check
               else
                 pool.wait_until_not_full
-                pool.wait_for_less_busy_worker(@options[:wait_for_less_busy_worker])
 
-                io = begin
-                  sock.accept_nonblock
-                rescue IO::WaitReadable
-                  next
+                # Wait for less busy worker?
+                if wait_for_less_busy_worker_enabled
+                  pool.wait_for_less_busy_worker(wait_for_less_busy_worker)
                 end
-                drain += 1 if shutting_down?
-                pool << Client.new(io, @binder.env(sock)).tap { |c|
-                  c.listener = sock
-                  c.send(addr_send_name, addr_value) if addr_value
-                }
+
+                io = \
+                  begin
+                    sock.accept_nonblock
+                  rescue IO::WaitReadable
+                    next
+                  end
+
+                # if 'drain_on_shutdown' is enabled:
+                # - increment drain counter when the server is shutting down
+                drain += 1 if drain && shutting_down?
+
+                # initialize new client
+                c = Client.new(io, binder.env(sock))
+
+                # set listener socket
+                c.listener = sock
+
+                # set client remote address
+                # Prefer using direct method/setter instead of calling `#send` method
+                case remote_address_option
+                when :value
+                  c.peerip = remote_address_value_option
+                when :header
+                  c.remote_addr_header = remote_address_header_option
+                when :proxy_protocol
+                  c.expect_proxy_proto = remote_address_proxy_protocol_option
+                end
```

Comment:
This case was originally outside of the while loop and used the following inside of it:
```ruby
c.send(addr_send_name, addr_value) if addr_value
```

Reply:
Hello @MSP-Greg 
I reverted my code, the case is now outside like it was before.

Cheers